qemu/qmp.go

```diff
@@ -530,55 +529,48 @@ func (q *QMP) mainLoop() {
 		close(q.disconnectedCh)
 	}()
 
-	var version []byte
 	var cmdDoneCh <-chan struct{}
+	var version *QMPVersion
+	ready := false
 
-DONE:
 	for {
-		var ok bool
 		select {
 		case cmd, ok := <-q.cmdCh:
 			if !ok {
 				return
 			}
 			_ = cmdQueue.PushBack(&cmd)
-		case version, ok = <-fromVMCh:
-			if !ok {
-				return
-			}
-			if cmdQueue.Len() >= 1 {
+
+			// We only want to execute the new cmd if QMP is
+			// ready and there are no other commands pending.
+			// If there are commands pending our new command
+			// will get run when the pending commands complete.
+			if ready && cmdQueue.Len() == 1 {
 				q.writeNextQMPCommand(cmdQueue)
 				cmdDoneCh = currentCommandDoneCh(cmdQueue)
 			}
-			break DONE
-		}
-	}
 
-	q.connectedCh <- q.parseVersion(version)
-
-	for {
-		select {
-		case cmd, ok := <-q.cmdCh:
+		case line, ok := <-fromVMCh:
 			if !ok {
 				return
 			}
-			_ = cmdQueue.PushBack(&cmd)
-
-			// We only want to execute the new cmd if there
-			// are no other commands pending.  If there are
-			// commands pending our new command will get
-			// run when the pending commands complete.
 
-			if cmdQueue.Len() == 1 {
-				q.writeNextQMPCommand(cmdQueue)
-				cmdDoneCh = currentCommandDoneCh(cmdQueue)
-			}
-		case line, ok := <-fromVMCh:
-			if !ok {
-				return
+			if !ready {
+				// Not ready yet. Check if line is the QMP version.
+				// Sometimes QMP events are thrown before the QMP version,
+				// hence it's not a guarantee that the first data read from
+				// the channel is the QMP version.
+				version = q.parseVersion(line)
+				if version != nil {
+					q.connectedCh <- version
+					ready = true
+					break
```

Comment:
I'm concerned this will lead to deadlock in the case that version is not nil.  We don't see this in the test case as we're calling startQMPLoop and not QMPStart

https://github.com/intel/govmm/blob/5944b68778fd11e3271c4904cfedd27e08d4b34e/qemu/qmp.go#L679

QMPStart waits for the version string before returning.  If the caller had supplied an eventCh and had not started his/her event loop until after QMPStart had returned successfully and events received before the version string would cause deadlock.  I think.  I haven't actually tested this.  We could fix this just by moving the break out of the if statement.  This would mean throwing the event away but I think that's fine.

Reply:
you're right :smile: